concordium-node/src/grpc2.rs

```diff
@@ -0,0 +1,526 @@
+                Err(e) => {
+                    error!("Could not acquire lock: {}", e);
+                    return Err(tonic::Status::internal("Could not enqueue request."));
+                }
+            }
+            Ok(tonic::Response::new(tokio_stream::wrappers::ReceiverStream::new(receiver)))
+        }
+
+        async fn get_account_info(
+            &self,
+            request: tonic::Request<crate::grpc2::types::AccountInfoRequest>,
+        ) -> Result<tonic::Response<Vec<u8>>, tonic::Status> {
+            if !self.service_config.get_account_info {
+                return Err(tonic::Status::unimplemented("`GetAccountInfo` is not enabled."));
+            }
+            let request = request.get_ref();
+            let block_hash = request.block_hash.require()?;
+            let account_identifier = request.account_identifier.require()?;
+            let (hash, response) =
+                self.consensus.get_account_info_v2(block_hash, account_identifier)?;
+            let mut response = tonic::Response::new(response);
+            add_hash(&mut response, hash)?;
+            Ok(response)
+        }
+
+        async fn get_account_list(
+            &self,
+            request: tonic::Request<crate::grpc2::types::BlockHashInput>,
+        ) -> Result<tonic::Response<Self::GetAccountListStream>, tonic::Status> {
+            if !self.service_config.get_account_list {
+                return Err(tonic::Status::unimplemented("`GetAccountList` is not enabled."));
+            }
+            let (sender, receiver) = futures::channel::mpsc::channel(100);
+            let hash = self.consensus.get_account_list_v2(request.get_ref(), sender)?;
+            let mut response = tonic::Response::new(receiver);
+            add_hash(&mut response, hash)?;
+            Ok(response)
+        }
+    }
+}
+
+/// Add hash to the provided response. The hash is added as part of metadata.
+fn add_hash<T>(response: &mut tonic::Response<T>, hash: [u8; 32]) -> Result<(), tonic::Status> {
+    let value = tonic::metadata::MetadataValue::try_from(hex::encode(&hash))
+        .map_err(|_| tonic::Status::internal("Cannot add metadata hash."))?;
+    response.metadata_mut().insert("blockhash", value);
+    Ok(())
+}
+
+/// A helper trait to make it simpler to require specific fields when parsing a
+/// protobuf message by allowing us to use method calling syntax and
+/// constructing responses that match the calling context, allowing us to use
+/// the `?` syntax.
+///
+/// The main reason for needing this is that in proto3 all non-primitive fields
+/// are optional, so it is up to the application to validate inputs if they are
+/// required.
+pub(crate) trait Require<E> {
+    type A;
+    fn require(&self) -> Result<&Self::A, E>;
+    /// Like [require](Self::require), but consumes the input. Note that
+    /// [require](Self::require) is not quite a special case of this, but for
+    /// the type `&A`. In particular [require](Self::require) ties the
```

Comment:
Something is off with this sentence.

Reply:
I have removed the sentence and the method `require_owned`. Turns out my reasoning was mostly bogus, and one can get good ergonomics with just the one method. The sacrifice is an occassional use of `.as_ref()`.